src/lib.rs

```diff
@@ -208,120 +238,225 @@ impl<'text> InitialInfoExt<'text> {
         text: &'a str,
         default_para_level: Option<Level>,
     ) -> InitialInfoExt<'a> {
-        let mut original_classes = Vec::with_capacity(text.len());
+        let (original_classes, paragraphs, pure_ltr) =
+            compute_initial_info(data_source, text, default_para_level);
+
+        InitialInfoExt {
+            base: InitialInfo {
+                text,
+                original_classes,
+                paragraphs,
+            },
+            pure_ltr,
+        }
+    }
+}
+
+/// Initial bidi information of the text (UTF-16 version).
+///
+/// Contains the text paragraphs and `BidiClass` of its characters.
+#[derive(PartialEq, Debug)]
+pub struct InitialInfoU16<'text> {
```

Comment:
thought: it would be ideal for us to have a GenericBidiInfo, GenericInitialInfo, and GenericParagraph types that abstract over an encoding, where the UTF8 one is just a type alias.

This might be more work. I'm fine with having separate UTF16 types for now as long as they're in a module.

Reply:
Yes, that'd be nice; my grasp of Rust is still pretty weak, though, so I haven't attempted this as yet.